includes/payment-methods/class-wc-stripe-upe-payment-method.php

```diff
@@ -128,16 +118,74 @@ public function is_reusable() {
 	}
 
 	/**
-	 * Add payment method to user and return WC payment token
+	 * Returns string representing payment method type
+	 * to query to retrieve saved payment methods from Stripe.
+	 */
+	public function get_retrievable_type() {
+		return $this->is_reusable() ? WC_Stripe_UPE_Payment_Method_Sepa::STRIPE_ID : null;
+	}
+
+	/**
+	 * Add payment method from intent to user and return WC payment token.
 	 *
-	 * @param WP_User $user User to get payment token from.
-	 * @param string  $payment_method_id Stripe payment method ID string.
+	 * By default we use WC_Payment_Token_SEPA, because most
+	 * payment methods support saving payment methods via
+	 * conversion to SEPA Direct Debit.
 	 *
-	 * @return WC_Payment_Token_CC|WC_Payment_Token_SEPA WC object for payment token.
+	 * @param WP_User $user User to add payment token to.
+	 * @param object $intent JSON object for Stripe payment intent.
+	 *
+	 * @return WC_Payment_Token_SEPA|null WC object for payment token.
 	 */
-	public function add_payment_token_to_user( $user, $payment_method_object ) {
-		$customer = new WC_Stripe_Customer( $user->ID );
-		return $customer->add_payment_method_to_user( $payment_method_object );
+	public function add_token_to_user_from_intent( $user_id, $intent ) {
+		if ( ! $this->is_reusable() ) {
+			return null;
+		}
+		$customer       = new WC_Stripe_Customer( $user_id );
+		$payment_method = $this->get_payment_method_details_from_intent( $intent );
+		$token          = $this->create_payment_token_for_user( $user_id, $payment_method->generated_sepa_debit, $payment_method->iban_last4 );
+
+		$customer->add_payment_method_actions( $token, $payment_method );
+		return $token;
+	}
+
+	/**
+	 * Add payment method to user and return WC payment token.
+	 *
+	 * This will be used from the WC_Stripe_Payment_Tokens service
+	 * as opposed to WC_Stripe_UPE_Payment_Gateway.
+	 *
+	 * @param string $user_id        WP_User ID
+	 * @param object $payment_method Stripe payment method object
+	 *
+	 * @return WC_Payment_Token_SEPA
+	 */
+	public function add_token_to_user_from_payment_method( $user_id, $payment_method ) {
+		$customer = new WC_Stripe_Customer( $user_id );
+		$token    = $this->create_payment_token_for_user( $user_id, $payment_method->id, $payment_method->sepa_debit->last4 );
+
+		$customer->add_payment_method_actions( $token, $payment_method );
+		return $token;
+	}
```

Comment:
This function has been modified slightly to support being called from the Account Add Payment Method screen via the UPE gateway.